docs: extract sections in how to write integration tests to their own howto guide (#2662)
canonical/operator: docs/howto/write-integration-tests-for-a-charm.md

Reviewer: I think it would be helpful to add some links to the feature-specific advice about integration tests. At minimum, for relations and configuration.

Jubilant provides several other helpers, in case you need to change the `wait` condition. See {external+jubilant:ref}`Use a custom wait condition <use_a_custom_wait_condition>`.

> For a Kubernetes charm, deploy with a resource. See this modified example (from {ref}`Manage resources <manage-resources-integration-tests>`):

Reviewer 2: Should this say "modified" when it's literally the same file?

Author: No it shouldn't. I have removed `modified` from this sentence. Thanks for the catch!

> For a Kubernetes charm, you will deploy it with a resource. See the modified example (from {ref}`Manage resources | Write integration tests <manage-resources-integration-tests>`):

Reviewer: A few small suggestions to make this more concise. I think we can keep the link pointing to the integration testing section without mentioning that in the link text.

```suggestion
For a Kubernetes charm, deploy with a resource. See this modified example (from {ref}`Manage resources <manage-resources-integration-tests>`):
```

> ### Deploy your charm with resources

Author: I am not sure if I should remove this heading.

The instructions for this section has been moved to the `Manage resources` page. Keeping this section here with one line

```
See: {ref}`Manage resources | Write integration tests <manage-resources-integration-tests>`
```

is confusing, I think.

Reviewer: I recommend removing the "Deploy your charm with resources" section from this page. I think it's worth mentioning resources in the "Deploy your charm" section. After the paragraph "Jubilant provides several other helpers [...]" we could say that for a Kubernetes charm you'll deploy with a resource, and show an example. Then provide the link to `manage-resources-integration-tests` for more general info about managing resources in integration tests.

Author: > we could say that for a Kubernetes charm you'll deploy with a resource, and show an example.

I have been thinking of an example for this, and came to a conclusion. `manage-resources-integration-tests` defines a conventional way to deploy with a resource integration test:

```python
import pathlib

import jubilant
import pytest
import yaml

METADATA = yaml.safe_load(pathlib.Path('./charmcraft.yaml').read_text())

@pytest.mark.juju_setup
def test_deploy(charm: pathlib.Path, juju: jubilant.Juju):
    resources = {
        name: res['upstream-source']
        for name, res in METADATA['resources'].items()
    }
    juju.deploy(charm, resources=resources)
    juju.wait(jubilant.all_active)
```

I think it's best to use the same example. `manage-resources-integration-tests` contains this example, and 2 short paragraph describing it.

I'm leaning toward duplicating this example here, and provide a link to `manage-resources-integration-tests`. Please let me know if this is a good idea.

Another way is to add this paragraph after `Jubilant provides several other helpers [...]`

```
Jubilant provides several other helpers [...]

For a Kubernetes charm, you will deploy it with a resource. See: {ref}`Manage resources | Write integration tests <manage-resources-integration-tests>` for more examples.
```

Reviewer: Hmm, I think this might be one time that duplication is the best approach. I can imagine someone searching or scanning the page looking for `upstream-source` or the general shape of this example. A link to `manage-resources-integration-tests` wouldn't make this possible.

You could consider putting the duplicated content in a separate file and reusing it. In general, I try to avoid that solution, as it makes the doc build more complicated. But I'd be OK with it here.

We already have a `reuse` directory: https://github.com/canonical/operator/tree/main/docs/reuse. You could put a `.md` file in there. See [how it's done for Pebble](https://github.com/canonical/pebble/blob/master/docs/how-to/install-pebble.md?plain=1#L46-L49). Although the `start-after` and `end-before` parts seem over complicated - I assume you wouldn't need them.

If you take this approach, it's best to exclude the `reuse` directory from the generated sitemap ([as for Pebble](https://github.com/canonical/pebble/blob/master/docs/conf.py#L136)).

Author: Thanks for the suggestion and guide. I thought about it and went with the `reuse` solution (in https://github.com/canonical/operator/pull/2662/commits/fab9460a13c06273539b9fe02d96b2b2e15a56f1)
